src/ipc/ipc4/handler.c

```diff
@@ -814,16 +806,30 @@ void ipc_cmd(ipc_cmd_hdr *_hdr)
 
 	/* FW sends a ipc message to host if request bit is set*/
 	if (in->r.rsp == SOF_IPC4_MESSAGE_DIR_MSG_REQUEST) {
+		char *data = ipc_get()->comp_data;
 		struct ipc4_message_reply reply;
-		struct sof_ipc_reply rep;
+		int try_count = 10;
+
+		while (msg_data.delayed_reply) {
```

Comment:
Whats teh use case here and while do we spin/wait ?

Reply:
Add comments. The schedule thread may be complete in 1ms, so spin wait may occupy CPU.

Reply:
sorry, I'm still not understanding why we need to busy wait here. Can you give a step by step example. I assume the wait is in preemptable context ?

Reply:
sure.  
(1) host send a ipc message to set pipeline status. A array of pipelines are included.
(2) ipc thread in FW call pipeline_trigger to schedule pipeline process in another threads , called schedule thread
(3) ipc thread waits for all these schedule thread complete and send back status
(4) ipc thread send status to host. Two types of error : (1) ipc thread gets a error when process ipc message (2) schedule thread reports a error. So we can't simply let ipc thread or schedule thread return error code. This patch lets ipc thread wait for all the schedule thread complete and get their status, then send back error to host.

Reply:
ok, so are we waiting on all pipeline states to be changed before we reply to host ?

What happens if one of the pipelines has an error during the state change ?
1) Do we stop other pipelines mentioned in the IPC message from changing state ?
2) Do we immediately reply with the error to the host and not wait for any other pipelines ?
3) What about the pipelines that successfully changed state, do we reset them ? (or does host do this)

I think it would probably be better to have a helper API for "compound" IPC messages where we have to perform multiple actions (where any could fail). e.g.
```c
/* tell the IPC logic we have a new compond IPC with num_messages */
int ipc_compound_start(struct ipc *ipc, int num_messages);

/* message ID completed OK  - decrements the reference count and replies if count is 0 */
void ipc_compound_msg_done(struct ipc *ipc, int message_id);

/* message ID completed error  - decrements the reference count and sets the reply error */
void ipc_compound_msg_error(struct ipc *ipc, int message_id, int error);
```

Reply:
> ok, so are we waiting on all pipeline states to be changed before we reply to host ?
> 
> What happens if one of the pipelines has an error during the state change ?
> 
> 1. Do we stop other pipelines mentioned in the IPC message from changing state ?

yes,  pipeline is processed one by one and stop to process next pipeline if error occurs


> 2. Do we immediately reply with the error to the host and not wait for any other pipelines ?

No,  ipc thread waits for all pipelines to be completed

> 3. What about the pipelines that successfully changed state, do we reset them ? (or does host do this)

Host does it

> I think it would probably be better to have a helper API for "compound" IPC messages where we have to perform multiple actions (where any could fail). e.g.
> 
> ```c
> /* tell the IPC logic we have a new compond IPC with num_messages */
> int ipc_compound_start(struct ipc *ipc, int num_messages);
> 
> /* message ID completed OK  - decrements the reference count and replies if count is 0 */
> void ipc_compound_msg_done(struct ipc *ipc, int message_id);
> 
> /* message ID completed error  - decrements the reference count and sets the reply error */
> void ipc_compound_msg_error(struct ipc *ipc, int message_id, int error);
> ```

Thanks. I made some refinement.

Reply:
> > ok, so are we waiting on all pipeline states to be changed before we reply to host ?
> > What happens if one of the pipelines has an error during the state change ?
> > 
> > 1. Do we stop other pipelines mentioned in the IPC message from changing state ?
> 
> yes, pipeline is processed one by one and stop to process next pipeline if error occurs
> 
> > 2. Do we immediately reply with the error to the host and not wait for any other pipelines ?
> 
> No, ipc thread waits for all pipelines to be completed
> 
> > 3. What about the pipelines that successfully changed state, do we reset them ? (or does host do this)
> 
> Host does it

ok, so host driver has to unwind a compound IPC message when one part fails. Is this needed for Windows compatibility ?

Btw, How is this conveyed to the host ? I assume the host driver knows which part of the compound IPC failed and which parts passed ?

@plbossart @ujfalusi @ranj063 fyi - may need some driver logic to track.

Reply:
Window will reset all the pipeline in last ipc messages since FW only replies status of the whole ipc message, no detail for each pipeline.

Window sends a batch of pipeline to fw in single ipc message but SOF Linux driver only sends a single pipeline in a single ipc message since we send it when visiting widget list.

Reply:
@RanderWang, I think this defies logic.
A. We have 1 pipeline in the array that we want to set the status
if the return value is **OK** then we **know** that the status of the pipeline has **been** changed
if the return value is **error** code then we **know** that the status of the pipeline has **not been** changed

B. We have more than 1 pipeline in the array that we want to set the status
if the return value is **OK** then we **know** that the status of the pipelines has **been** changed
if the return value is **error** code then we **don't know** the status of the pipelines

In case of B and error any of the pipelines or all of them could have failed with the status change, right?
Is it an error if some went without error, but some failed?
If it is an error then why would you continue after a failed status change?
If there is a status change failure why are we not rolling back the change we did successfully prior?

imho the rollback has to be done at the place of failure.

This logic is like:
```c
ret = set_up_few_things(dev, number_of_things);
if (ret ~= 0)
	release_the_things(dev, number_of_things);
```
When `set_up_few_things()` fails it is not expected that there is anything left to clean up as it has returned with error, it has failed.

Reply:
````
If it is an error then why would you continue after a failed status change?
If there is a status change failure why are we not rolling back the change we did successfully prior?
````
we stop pipeline process when we get a error in ipc thread. If the error happened in schedule thread, we  don't wait for it but will get the error code at end and return to host and host will do the reset not fw.